src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala

```diff
@@ -361,20 +358,78 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
             tp.typeSymbol.owner == tp.prefix.typeSymbol && isStatic(tp.prefix)
           }
 
-          if ((expectedPrefix eq NoPrefix)
-            || expectedTp.typeSymbol.isJava
-            || definedInStaticLocation(expectedTp)
-            || testedPrefix =:= expectedPrefix) orig
-          else gen.mkAttributedQualifierIfPossible(expectedPrefix) match {
-            case None => orig
-            case Some(expectedOuterRef) =>
-              // ExplicitOuter replaces `Select(q, outerSym) OBJ_EQ expectedPrefix`
-              // by `Select(q, outerAccessor(outerSym.owner)) OBJ_EQ expectedPrefix`
-              // if there's an outer accessor, otherwise the condition becomes `true`
-              // TODO: centralize logic whether there's an outer accessor and use here?
-              val synthOuterGetter = expectedTp.typeSymbol.newMethod(vpmName.outer, newFlags = SYNTHETIC | ARTIFACT) setInfo expectedPrefix
-              val outerTest = (Select(codegen._asInstanceOf(testedBinder, expectedTp), synthOuterGetter)) OBJ_EQ expectedOuterRef
-              and(orig, outerTest)
+          // In `def foo(a: b.B) match { case _: p.P }`
+          // testedBinder.symbol.info = b.B
+          // expectedTp                = p.P
+
+          expectedTp match {
+            case _ if expectedTp.typeSymbolDirect.isAliasType =>
+              withOuterTest(orig)(testedBinder, expectedTp.dealias)
```

Comment:
Are the usages of `expectedTp` fine with no longer using the original alias?

Reply:
This dealias is needed to have the behaviour in the cases below.

```
class Outer { class A; class B; type T = A with B }
val o = new Outer
type T1 = o.A with o.B
def pat1(a: Any) = a match { case o.T => }
def pat2(a: Any) = a match { case T1 => }
def pat3(a: Any) = a match { case o.A with o.A => }
```

I can't see any assumptions in the other usages of `expectedTp` that would be violated by recursing into the alias.